nomos-libp2p/src/behaviour/nat/inner.rs

```diff
@@ -269,3 +269,207 @@ where
         Poll::Pending
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use std::{cell::RefCell, pin::Pin};
+
+    use libp2p::{
+        autonat::v2::server,
+        core::{
+            transport::{DialOpts, ListenerId, MemoryTransport, Transport, TransportError},
+            upgrade::Version,
+        },
+        identify,
+        identity::Keypair,
+        plaintext,
+        swarm::{NetworkBehaviour, SwarmEvent},
+        yamux,
+    };
+    use libp2p_swarm_test::SwarmExt as _;
+    use rand::{
+        rngs::{OsRng, StdRng},
+        SeedableRng as _,
+    };
+
+    use super::*;
+    use crate::{
+        behaviour::nat::address_mapper::errors::AddressMapperError, config::NatMappingSettings,
+    };
+
+    thread_local! {
+        static MAPPER_CALLED: RefCell<bool> = const { RefCell::new(false) };
+    }
+
+    #[tokio::test]
+    async fn test_nat_mapping_happy_flow() {
+        let (mut server, server_addr) = create_server_with_blocking_transport();
+        let (mut client, expected_addr) = create_mapped_client();
+
+        client.dial(server_addr).unwrap();
+
+        tokio::time::timeout(Duration::from_secs(30), async {
+            loop {
+                tokio::select! {
+                    event = client.next_swarm_event() => {
+                        if let SwarmEvent::ExternalAddrConfirmed { address } = event {
+                            if address == expected_addr {
+                                return;
+                            }
+                        }
+                    },
+                    _ = server.next_swarm_event() => {}
+                }
+            }
+        })
+        .await
+        .unwrap();
+
+        assert!(
+            MAPPER_CALLED.with(|called| *called.borrow()),
+            "Mapper should have been called"
+        );
+    }
+
+    struct TestMapper;
+
+    #[async_trait::async_trait]
+    impl NatMapper for TestMapper {
+        async fn map_address(
+            _address_to_map: &Multiaddr,
+            _settings: NatMappingSettings,
+        ) -> Result<Multiaddr, AddressMapperError> {
+            MAPPER_CALLED.with(|called| *called.borrow_mut() = true);
+            Ok("/memory/2000".parse().unwrap())
```

Comment:
Could we choose a random port to avoid concurrency or re-use issues in the future?